packages/@lwc/engine/src/framework/restrictions.ts

```diff
@@ -382,12 +387,49 @@ function getComponentRestrictionsDescriptors(): PropertyDescriptorMap {
     };
 }
 
-function getLightningElementPrototypeRestrictionsDescriptors(proto: object): PropertyDescriptorMap {
+function getLightningElementPrototypeRestrictionsDescriptors(
+    proto: LightningElement
+): PropertyDescriptorMap {
     if (process.env.NODE_ENV === 'production') {
         // this method should never leak to prod
         throw new ReferenceError();
     }
-    const descriptors = {};
+
+    const originalDispatchEvent = proto.dispatchEvent;
+
+    const descriptors = {
+        dispatchEvent: generateDataDescriptor({
+            value(this: LightningElement, event: Event): boolean {
+                const vm = getComponentVM(this);
+
+                assert.isFalse(
+                    isBeingConstructed(vm),
+                    `this.dispatchEvent() should not be called during the construction of the custom element for ${getComponentAsString(
+                        this
+                    )} because no one is listening just yet.`
+                );
+
+                if (isObject(event)) {
+                    const { type } = event;
+
+                    if (isString(type) && !/^[a-z][a-z0-9_]*$/.test(type)) {
+                        logError(
+                            `Invalid event type "${type}" dispatched in element ${getComponentAsString(
+                                vm.component
+                            )}. ` +
+                                'Event name must start with a lowercase letter and followed only lowercase letters, numbers, and underscores',
+                            vm.elm
+                        );
+                    }
+                }
+
+                // Typescript does not like it when you treat the `arguments` object as an array
+                // @ts-ignore type-mismatch
+                return originalDispatchEvent.apply(this, arguments);
```

Comment:
slice the arguments

Reply:
See other comment related to array slicing.